orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/StartExecutionHandler.kt

```diff
@@ -40,36 +42,57 @@ class StartExecutionHandler(
 
-  val log: Logger
-    get() = LoggerFactory.getLogger(javaClass)
+  private val log: Logger get() = LoggerFactory.getLogger(javaClass)
 
   override fun handle(message: StartExecution) {
     message.withExecution { execution ->
       if (execution.status == NOT_STARTED && !execution.isCanceled) {
-        val initialStages = execution.initialStages()
-        if (initialStages.isEmpty()) {
-          log.warn("No initial stages found (executionId: ${message.executionId})")
-          repository.updateStatus(message.executionId, TERMINAL)
-          publisher.publishEvent(ExecutionComplete(this, message.executionType, message.executionId, TERMINAL))
-          return@withExecution
-        }
-
-        repository.updateStatus(message.executionId, RUNNING)
-        initialStages
-          .forEach {
-            queue.push(StartStage(message, it.id))
-          }
-        publisher.publishEvent(ExecutionStarted(this, message.executionType, message.executionId))
-      } else {
-        if (execution.status == CANCELED || execution.isCanceled) {
-          publisher.publishEvent(ExecutionComplete(this, message.executionType, message.executionId, execution.status))
+        if (execution.shouldQueue()) {
+          log.info("Queueing {} {} {}", execution.application, execution.name, execution.id)
         } else {
-          log.warn("Execution (type: ${message.executionType}, id: {}, status: ${execution.status}, application: {})" +
-            " cannot be started unless state is NOT_STARTED. Ignoring StartExecution message.",
-            value("executionId", message.executionId),
-            value("application", message.application))
+          start(execution)
         }
+      } else {
+        terminate(execution)
       }
     }
   }
+
+  private fun start(execution: Execution) {
+    val initialStages = execution.initialStages()
+    if (initialStages.isEmpty()) {
+      log.warn("No initial stages found (executionId: ${execution.id})")
+      repository.updateStatus(execution.id, TERMINAL)
+      publisher.publishEvent(ExecutionComplete(this, execution.type, execution.id, TERMINAL))
+    } else {
+      repository.updateStatus(execution.id, RUNNING)
+      initialStages.forEach { queue.push(StartStage(it)) }
+      publisher.publishEvent(ExecutionStarted(this, execution.type, execution.id))
+    }
+  }
+
+  private fun terminate(execution: Execution) {
+    if (execution.status == CANCELED || execution.isCanceled) {
+      publisher.publishEvent(ExecutionComplete(this, execution.type, execution.id, execution.status))
+    } else {
+      log.warn("Execution (type: ${execution.type}, id: {}, status: ${execution.status}, application: {})" +
+        " cannot be started unless state is NOT_STARTED. Ignoring StartExecution message.",
+        value("executionId", execution.id),
+        value("application", execution.application))
+    }
+  }
+
+  private fun Execution.shouldQueue(): Boolean {
```

Comment:
Queuing logic moved into `StartExecutionHandler`. We can simply no-op if there's already an instance of this pipeline running.